crates/uv-python/fetch-download-metadata.py

```diff
@@ -1,272 +1,389 @@
-    % (
-        "|".join(
-            map(
-                re.escape,
-                sorted(FLAVOR_PREFERENCES + HIDDEN_FLAVORS, key=len, reverse=True),
-            )
-        )
-    )
-)
-
-# Normalized mappings to match the Rust types
-ARCH_MAP = {
-    "ppc64": "powerpc64",
-    "ppc64le": "powerpc64le",
-}
-
-
-def parse_filename(filename):
-    match = _filename_re.match(filename)
-    if match is None:
-        return
-    version, triple = match.groups()
-    if triple.endswith("-full"):
-        triple = triple[:-5]
-    match = _flavor_re.match(triple)
-    if match is not None:
-        triple, flavor = match.groups()
-    else:
-        flavor = None
-
-    return (version, triple, flavor)
-
-
-def normalize_triple(triple):
-    if "-static" in triple:
-        logging.debug("Skipping %r: static unsupported", triple)
-        return
-    triple = SPECIAL_TRIPLES.get(triple, triple)
-    pieces = triple.split("-")
-    try:
-        arch = normalize_arch(pieces[0])
-        operating_system = normalize_os(pieces[2])
-        if pieces[2] == "linux":
-            # On linux, the triple has four segments, the last one is the libc
-            libc = pieces[3]
-        else:
-            libc = "none"
-    except IndexError:
-        logging.debug("Skipping %r: unknown triple", triple)
-        return
-    return "%s-%s-%s" % (arch, operating_system, libc)
 
 
-def normalize_arch(arch):
-    arch = ARCH_MAP.get(arch, arch)
-    pieces = arch.split("_")
-    # Strip `_vN` from `x86_64`
-    return "_".join(pieces[:2])
+def batched(iterable: Iterable, n: int) -> Generator[tuple, None, None]:
+    """Batch data into tuples of length n. The last batch may be shorter."""
+    # batched('ABCDEFG', 3) --> ABC DEF G
+    if n < 1:
+        raise ValueError("n must be at least one")
+    it = iter(iterable)
+    while batch := tuple(itertools.islice(it, n)):
+        yield batch
 
 
-def normalize_os(os):
-    return os
+class PlatformTriple(NamedTuple):
+    arch: str
+    platform: str
+    libc: str
 
 
-def read_sha256(url):
-    try:
-        resp = request(url + ".sha256")
-    except urllib.error.HTTPError:
-        return None
-    assert resp.status == 200
-    return resp.read().decode().strip()
+class Version(NamedTuple):
+    major: int
+    minor: int
+    patch: int
 
+    @classmethod
+    def from_str(cls, version: str) -> Self:
+        major, minor, patch = version.split(".", 3)
+        return cls(int(major), int(minor), int(patch))
 
-def sha256(path):
-    h = hashlib.sha256()
+    def __str__(self) -> str:
+        return f"{self.major}.{self.minor}.{self.patch}"
 
-    with open(path, "rb") as file:
-        while True:
-            # Reading is buffered, so we can read smaller chunks.
-            chunk = file.read(h.block_size)
-            if not chunk:
-                break
-            h.update(chunk)
+    def __neg__(self) -> Self:
+        return Version(-self.major, -self.minor, -self.patch)
```

Comment:
This feels a little weird. Is this actually the easiest way to reverse the versions?